Std/Lean/Meta/Basic.lean

```diff
@@ -274,6 +274,20 @@ where
           modify (·.insert pendingMVarId)
         go pendingMVarId
 
+/-- Try to close the goal using `refl`. If `refl` fails return the `mvarId` unchanged and
+return `none` otherwise.
+
+This formulation is especially useful to `filterMapM` a list of goals. -/
+def notRefl? (mvarId : MVarId) : MetaM (Option MVarId) := do
```

Comment:
I think this one would be better off as `MetaM Bool` reporting success or not, since it just returns the input unchanged on failure, but at that point it's not much better than just open-coding the whole thing. I would like to see the use case for this definition.

Also the name `notRefl?` is pretty bad, `tryRefl?` or even `refl?` is better.

Reply:
My use-case is that I want to close all goals that can be proved by `refl`. In my case I just want to filter through a list of metavariables, but one could also think of adding an option for Kyle's `congr!` so that it only uses `refl` but not `assumption`, see https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Tactic/Congr!.lean#L818
I suspect this to be more convenient in the negated form, that is if `refl` succeeds then return `false` so that you can `filterM` a list (or `filterMapM` in the current version). But you have way more experience in this stuff, so if you prefer that reports success then that is fine with me as well.

Reply:
I think the bool-valued one that returns whether it succeeds is the one that would fit in the best with the rest of the Lean library.
```
def refl? (mvarId : MVarId) : MetaM Bool := do
  try mvarId.refl; return true catch _ => return false
```
I'm not sure it's necessary to use `observing?` unless `isDefEq` has any side effects when it fails.

For your use case, it's not so bad to negate the result to filter out when it succeeds:
```
def mapRefl (mvarIds : List MVarId) : MetaM (List MVarId) :=
  mvarIds.filterM (not <$> refl? ·)
```

Reply:
Don't have a strong opinion about the fate of this function, but noting the related `Lean.MVarId.rfl : MVarId \to MetaM Unit` in [`Mathlib.Tactic.Relation.Rfl`](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Tactic/Relation/Rfl.lean#L49).